pkg/anchor/parser.go

```diff
@@ -0,0 +1,146 @@
+package anchor
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+	"unicode"
+)
+
+type commandType int
+
+const (
+	CommandFrom commandType = iota
+	CommandRun
+	CommandOther
+)
+
+type EntryType int
+
+const (
+	EntryCommand EntryType = iota
+	EntryComment
+	EntryEmpty
+)
+
+type Entry struct {
+	Type      EntryType
+	Value     string
+	Beginning bool
+}
+
+type Nodes []Node
+
+type Node struct {
+	Entries     []Entry
+	CommandType commandType
+	Command     string
+}
+
+func (n Nodes) Print() {
+	for _, node := range n {
+		for _, entry := range node.Entries {
+			fmt.Println(entry.Value)
+		}
+	}
+}
+
+func (n Nodes) Write(w io.Writer) {
+	for _, node := range n {
+		node.Write(w)
+	}
+}
+
+func (n Node) Write(w io.Writer) {
+	b := []byte{}
+	for _, entry := range n.Entries {
+		b = append(b, []byte(entry.Value)...)
+	}
+	w.Write(b)
+}
+
+func (n *Node) appendLine(line []byte, entryType EntryType, beginning bool, strip ...string) {
+	if entryType == EntryCommand {
+		l := string(line)
+		for _, s := range strip {
+			l = strings.TrimLeft(l, s)
+			l = strings.TrimSpace(l)
+		}
+		n.Command += l
+	}
+
+	// new lines are trimmed by the scanner so we re-add them here
+	line = append(line, '\n')
+	n.Entries = append(n.Entries, Entry{Type: entryType, Value: string(line), Beginning: beginning})
+}
+
+func Parse(r io.Reader) Nodes {
+	scanner := bufio.NewScanner(r)
+	node := Node{}
+	nodes := make([]Node, 0)
+	for scanner.Scan() {
+		line := scanner.Bytes()
+
+		if isComment(line) {
```

Comment:
this `ifelse` chain _could_ be a switch statement, but I'm okay if it stays as-in